### PR TITLE
Os 386 cancellation tokens

### PR DESCRIPTION
**Pull request by jared-improbable**

### CODE REVIEW

File: Improbable/Stdlib/Improbable.Stdlib.CSharpCodeGen/ProjectInfoGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+using static Improbable.CSharpCodeGen.Case;
+
+namespace Improbable.Stdlib.CSharpCodeGen
+{
+    public class ProjectInfoGenerator
+    {
+        public string GenerateProjectInfo(string pathToSpatialOsJson, string ns = "SpatialOsProject")
+        {
+            var project = JsonConvert.DeserializeObject<ProjectConfig>(File.ReadAllText(pathToSpatialOsJson, Encoding.UTF8));
+            var configDir = Path.GetDirectoryName(pathToSpatialOsJson) ?? Environment.CurrentDirectory;
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib.CSharpCodeGen/ProjectInfoGenerator.cs]: So `Path.GetDirectoryName` only returns null for if the directory is root (or is null). 

In these cases using `Environment.CurrentDirectory` feels wrong?

File: Improbable/Stdlib/Improbable.Stdlib.CSharpCodeGen/ProjectInfoGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+            var layers = new HashSet<string>();
+            var workerTypes = new HashSet<string>();
+            var mapping = new Dictionary<string, List<string>>();
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib.CSharpCodeGen/ProjectInfoGenerator.cs]: mapping of... what?

File: Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs [hunk 3]
```diff
@@ -43,41 +44,59 @@ public string WorkerId
...
                 {
                     using (var future = Connection.ConnectAsync(host, port, workerName, connectionParameters))
                     {
-                        var connection = future.Get();
+                        Connection connection;
+                        while (true)
+                        {
+                            cancellation.ThrowIfCancellationRequested();
+
+                            if (future.TryGet(out connection, 50))
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs]: Ugh this "future" should _really_ be a `Task<Connection>` and this polling should be pushed back into the SDK imo

File: Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs [hunk 4]
```diff
@@ -125,9 +148,18 @@ private static Task<WorkerConnection> ConnectAsync(ILocatorOptions options, Conn
                     using (var locator = new Locator(options.SpatialOsHost, options.SpatialOsPort, locatorParameters))
                     using (var future = locator.ConnectAsync(connectionParameters))
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs]: Some common code for polling the future that can be factored out. 


We might actually be able to write a generic extension method to transform it into a task.


```csharp
        public static Task<T> ToTask<T>(this Future<T> future, CancellationToken cancellation = default)
        {
            var tcs = new TaskCompletionSource<T>();
            cancellation.Register(() => tcs.TrySetCanceled(cancellation));

            Task.Run(() =>
            {
                while (true)
                {
                    cancellation.ThrowIfCancellationRequested();

                    if (future.TryGet(out var value, 50))
                    {
                        tcs.TrySetResult(value);
                        break;
                    }
                }
            }, cancellation);

            return tcs.Task;
        }
```

Or something like that

File: Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs [hunk 5]
```diff
@@ -535,15 +561,11 @@ public OpList GetOpList(TimeSpan timeout)
...
             if (connection == null || GetConnectionStatusCode() != ConnectionStatusCode.Success)
             {
-                if (GetConnectionStatusCode() != ConnectionStatusCode.Success)
-                {
-                    CancelCommands();
-                }
-
-                return EmptyOpList;
+                CancelCommands();
+                throw new Exception("Not connected to SpatialOS");
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs]: `InvalidOperationException`?

File: Improbable/Test/Improbable.Test/ProcessExtensions.cs [hunk 6]
```diff
@@ -1,125 +0,0 @@
-using System;
```
Comment (jamiebrynes7) [inline Improbable/Test/Improbable.Test/ProcessExtensions.cs:1]: 🎉

File: Improbable/Stdlib/Improbable.Stdlib.CSharpCodeGen/ProjectInfoGenerator.cs (hunk 2 above)
Comment (jared-improbable) [reply Improbable/Stdlib/Improbable.Stdlib.CSharpCodeGen/ProjectInfoGenerator.cs]: Oops, leftover from a previous experiment - this is unneeded now.

File: Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs (hunk 4 above)
Comment (jared-improbable) [reply Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs]: Nice, good idea

File: Improbable/Stdlib/Improbable.Stdlib/EventProcessor.cs [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Improbable.Worker.CInterop;
+
+namespace Improbable.Stdlib
+{
+    public class EventProcessor<T> : IOpProcessor where T : struct
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib/EventProcessor.cs:11]: Is this the most specific generic constraint?

File: Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs [hunk 8]
```diff
@@ -88,41 +90,61 @@ public static async Task<Deployment> StartLocalAsync(StartDeploymentOptions star
...
-            var snapshotFile = Path.Combine(Path.GetDirectoryName(startDeployment.ProjectConfigPath), "snapshots", $"{startDeployment.SnapshotId}.snapshot");
+            var snapshotFile = Path.Combine(Path.GetDirectoryName(startDeployment.ProjectConfigPath) ?? "", "snapshots", $"{startDeployment.SnapshotId}.snapshot");
             if (!File.Exists(snapshotFile))
             {
                 throw new FileNotFoundException(snapshotFile);
             }
 
-            progress?.Report($"Starting local deployment for '{project.ProjectName}' with snapshot '{startDeployment.SnapshotId}'...");
-            var response = client.CreateDeployment(new CreateDeploymentRequest
+            var retries = 10;
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs:104]: Nit: can be const

**Review by jamiebrynes7 — APPROVED** [review]

File: Improbable/Stdlib/Improbable.Stdlib/EventProcessor.cs (hunk 7 above)
Comment (jared-improbable) [reply Improbable/Stdlib/Improbable.Stdlib/EventProcessor.cs:11]: Yeah, there's no inheritance or anything else to go against

**Pull request merged**